File: csrc/sageattn/ops.cpp
/*
 * Copyright (c) 2024 by SageAttention team.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// TORCH_LIBRARY registration for torch.ops.sageattention.*.
//
// Layers:
//   - 8 low-level quantization ops (mutating, -> (), no fake needed): the
//     former pybind11 _fused surface, overloads merged via optional args.
//   - host query ops: plan (dispatch decision table) and compiled_archs.
//
// This TU is host-only: it sees the fused.h declarations; kernels live in
// their own TUs. The attention ops (fwd / fwd_varlen) register in theirs.

#include <optional>
#include <string>
#include <vector>

#include <torch/library.h>
#include <torch/types.h>

#include "../fused/fused.h"
#include "config.h"
#include "plan.h"
#include "sageattn_build_config.h"
#include "varlen_check.h"

namespace sage {
namespace {

// ---------------------------------------------------------------- helpers

inline int layout_flag(c10::string_view s)
{
    return static_cast<int>(parse_tensor_layout(s));
}

// ---------------------------------------------------------------- fused ops

void quant_per_block_int8(const at::Tensor&                input,
                          const std::optional<at::Tensor>& mean,
                          at::Tensor&                      output,
                          at::Tensor&                      scale,
                          std::optional<double>            sm_scale,
                          int64_t                          block_size,
                          c10::string_view                 tensor_layout)
{
    const int layout_int = layout_flag(tensor_layout);
    if (mean.has_value()) {
        TORCH_CHECK(!sm_scale.has_value(),
                    "quant_per_block_int8: sm_scale and mean are mutually exclusive "
                    "(the fused sub-mean path has no sm_scale)");
        quant_per_block_int8_fuse_sub_mean_cuda(input, *mean, output, scale, static_cast<int>(block_size), layout_int);
    }
    else if (sm_scale.has_value()) {
        quant_per_block_int8_cuda(
            input, output, scale, static_cast<float>(*sm_scale), static_cast<int>(block_size), layout_int);
    }
    else {
        quant_per_block_int8_cuda(input, output, scale, static_cast<int>(block_size), layout_int);
    }
}

void quant_per_warp_int8(const at::Tensor& input,
                         at::Tensor&       output,
                         at::Tensor&       scale,
                         int64_t           block_size,
                         int64_t           warp_block_size,
                         c10::string_view  tensor_layout)
{
    quant_per_warp_int8_cuda(input,
                             output,
                             scale,
                             static_cast<int>(block_size),
                             static_cast<int>(warp_block_size),
                             layout_flag(tensor_layout));
}

void quant_per_thread_int8_q(const at::Tensor& input,
                             at::Tensor&       output,
                             at::Tensor&       scale,
                             int64_t           block_size,
                             int64_t           warp_block_size,
                             c10::string_view  tensor_layout)
{
    quant_per_thread_int8_q_cuda(input,
                                 output,
                                 scale,
                                 static_cast<int>(block_size),
                                 static_cast<int>(warp_block_size),
                                 layout_flag(tensor_layout));
}

void quant_per_thread_int8_k(const at::Tensor&                input,
                             const std::optional<at::Tensor>& mean,
                             at::Tensor&                      output,
                             at::Tensor&                      scale,
                             int64_t                          block_size,
                             int64_t                          warp_block_size,
                             c10::string_view                 tensor_layout)
{
    const int layout_int = layout_flag(tensor_layout);
    if (mean.has_value()) {
        quant_per_thread_int8_k_fuse_sub_mean_cuda(
            input, *mean, output, scale, static_cast<int>(block_size), static_cast<int>(warp_block_size), layout_int);
    }
    else {
        quant_per_thread_int8_k_cuda(
            input, output, scale, static_cast<int>(block_size), static_cast<int>(warp_block_size), layout_int);
    }
}

void sub_mean(const at::Tensor& input, const at::Tensor& mean, at::Tensor& output, c10::string_view tensor_layout)
{
    sub_mean_cuda(input, mean, output, layout_flag(tensor_layout));
}

// cu_seqlens switches the packed layout on: 3-D [total_tokens, heads,
// head_dim] in, [heads, head_dim, blk_total * pad_multiple] out. It is the one
// piece of the fp8 V pipeline that is pure fp16 arithmetic, so it is also the
// only piece whose packed addressing can be checked on a pre-sm_89 device -
// hence the argument here and not only inside quant_v_fp8_varlen.
void transpose_pad_v(const at::Tensor&                input,
                     at::Tensor&                      output,
                     c10::string_view                 tensor_layout,
                     bool                             permute,
                     const std::optional<at::Tensor>& cu_seqlens,
                     c10::SymInt                      max_seqlen,
                     int64_t                          pad_multiple)
{
    QuantVarlen varlen;
    if (cu_seqlens.has_value()) {
        check_cu_seqlens(*cu_seqlens, input, "cu_seqlens");
        varlen.cu_seqlens = cu_seqlens->data_ptr<int32_t>();
        varlen.batch_size = cu_seqlens->size(0) - 1;
        varlen.max_seqlen = max_seqlen.guard_int(__FILE__, __LINE__);
        varlen.pad_tokens = pad_multiple;
    }
    if (permute) {
        transpose_pad_permute_cuda(input, output, layout_flag(tensor_layout), varlen);
    }
    else {
        transpose_pad_cuda(input, output, layout_flag(tensor_layout), varlen);
    }
}

void scale_fuse_quant(const at::Tensor& input,
                      at::Tensor&       output,
                      at::Tensor&       scale,
                      int64_t           num_tokens,
                      double            scale_max,
                      c10::string_view  tensor_layout)
{
    scale_fuse_quant_cuda(
        input, output, scale, static_cast<int>(num_tokens), static_cast<float>(scale_max), layout_flag(tensor_layout));
}

void mean_scale_fuse_quant(const at::Tensor& input,
                           at::Tensor&       output,
                           at::Tensor&       mean,
                           at::Tensor&       scale,
                           int64_t           num_tokens,
                           double            scale_max,
                           c10::string_view  tensor_layout)
{
    mean_scale_fuse_quant_cuda(input,
                               output,
                               mean,
                               scale,
                               static_cast<int>(num_tokens),
                               static_cast<float>(scale_max),
                               layout_flag(tensor_layout));
}

// ---------------------------------------------------------------- host ops

std::vector<int64_t> compiled_archs()
{
    std::vector<int64_t> archs;
#if SAGEATTN_BUILD_SM80
    archs.push_back(80);
#endif
#if SAGEATTN_BUILD_SM89
    archs.push_back(89);
#endif
#if SAGEATTN_BUILD_SM90
    archs.push_back(90);
#endif
#if SAGEATTN_BUILD_SM100
    archs.push_back(100);
#endif
#if SAGEATTN_BUILD_SM120
    archs.push_back(120);
#endif
    return archs;
}

// Full dispatch decision for one (cc, head_dim, requests) tuple. Pure host —
// primed into a Python dict at import time (returning str from inside a
// traced region breaks fullgraph, so Python never calls this in hot code).
std::tuple<std::string,
           std::string,
           std::string,
           bool,
           bool,
           bool,
           std::string,
           int64_t,
           double,
           bool,
           bool,
           int64_t,
           int64_t,
           int64_t,
           int64_t,
           std::string>
plan(int64_t                         cc_major,
     int64_t                         cc_minor,
     int64_t                         head_dim,
     std::optional<c10::string_view> backend,
     std::optional<c10::string_view> qk_quant_gran,
     std::optional<c10::string_view> pv_accum_dtype,
     std::optional<bool>             smooth_v,
     bool                            varlen)
{
    std::optional<Backend>   req_backend;
    std::optional<QuantGran> req_gran;
    std::optional<PVAccum>   req_pv;
    if (backend)
        req_backend = parse_backend(*backend);
    if (qk_quant_gran)
        req_gran = parse_quant_gran(*qk_quant_gran);
    if (pv_accum_dtype)
        req_pv = parse_pv_accum(*pv_accum_dtype);

    Plan p = resolve(CC{static_cast<int>(cc_major), static_cast<int>(cc_minor)},
                     static_cast<int>(head_dim),
                     req_backend,
                     req_gran,
                     req_pv,
                     smooth_v,
                     varlen);
    return {name(p.backend),
            name(p.gran),
            name(p.pv),
            p.smooth_v,
            p.smooth_v_ignored,
            p.pv_fp8,
            name(p.v_layout),
            p.v_pad_multiple,
            p.v_scale_max,
            p.need_value_scale,
            p.need_value_mean,
            p.blk_q,
            p.warp_q,
            p.blk_k,
            p.warp_k,
            p.error};
}

}  // namespace
}  // namespace sage

// ---------------------------------------------------------------- registry

TORCH_LIBRARY(sageattention, m)
{
    // host queries (catch-all: no tensors involved)
    m.def("compiled_archs() -> int[]", TORCH_FN(sage::compiled_archs));
    m.def("plan(int cc_major, int cc_minor, int head_dim, str? backend=None, "
          "str? qk_quant_gran=None, str? pv_accum_dtype=None, bool? smooth_v=None, "
          "bool varlen=False) "
          "-> (str backend, str qk_quant_gran, str pv_accum_dtype, bool smooth_v, "
          "bool smooth_v_ignored, bool pv_fp8, str v_layout, int v_pad_multiple, "
          "float v_scale_max, bool need_value_scale, bool need_value_mean, "
          "int blk_q, int warp_q, int blk_k, int warp_k, str error)",
          TORCH_FN(sage::plan));

    // low-level quantization (mutating, no fake kernels needed)
    m.def("quant_per_block_int8(Tensor input, Tensor? mean, Tensor(a!) output, "
          "Tensor(b!) scale, float? sm_scale, int block_size, str tensor_layout) -> ()");
    m.def("quant_per_warp_int8(Tensor input, Tensor(a!) output, Tensor(b!) scale, "
          "int block_size, int warp_block_size, str tensor_layout) -> ()");
    m.def("quant_per_thread_int8_q(Tensor input, Tensor(a!) output, Tensor(b!) scale, "
          "int block_size, int warp_block_size, str tensor_layout) -> ()");
    m.def("quant_per_thread_int8_k(Tensor input, Tensor? mean, Tensor(a!) output, "
          "Tensor(b!) scale, int block_size, int warp_block_size, str tensor_layout) -> ()");
    m.def("sub_mean(Tensor input, Tensor mean, Tensor(a!) output, str tensor_layout) -> ()");
    m.def("transpose_pad_v(Tensor input, Tensor(a!) output, str tensor_layout, "
          "bool permute, Tensor? cu_seqlens=None, *, SymInt max_seqlen=0, "
          "int pad_multiple=64) -> ()");
    m.def("scale_fuse_quant(Tensor input, Tensor(a!) output, Tensor(b!) scale, "
          "int num_tokens, float scale_max, str tensor_layout) -> ()");
    m.def("mean_scale_fuse_quant(Tensor input, Tensor(a!) output, Tensor(b!) mean, "
          "Tensor(c!) scale, int num_tokens, float scale_max, str tensor_layout) -> ()");
}

TORCH_LIBRARY_IMPL(sageattention, CUDA, m)
{
    m.impl("quant_per_block_int8", TORCH_FN(sage::quant_per_block_int8));
    m.impl("quant_per_warp_int8", TORCH_FN(sage::quant_per_warp_int8));
    m.impl("quant_per_thread_int8_q", TORCH_FN(sage::quant_per_thread_int8_q));
    m.impl("quant_per_thread_int8_k", TORCH_FN(sage::quant_per_thread_int8_k));
    m.impl("sub_mean", TORCH_FN(sage::sub_mean));
    m.impl("transpose_pad_v", TORCH_FN(sage::transpose_pad_v));
    m.impl("scale_fuse_quant", TORCH_FN(sage::scale_fuse_quant));
    m.impl("mean_scale_fuse_quant", TORCH_FN(sage::mean_scale_fuse_quant));
}

// plan / compiled_archs need a CompositeExplicitAutograd (catch-all) impl —
// they take no tensors, so the CUDA dispatch key never matches.
TORCH_LIBRARY_IMPL(sageattention, CompositeExplicitAutograd, m)
{
    // registered inline in TORCH_LIBRARY above via m.def(schema, fn)
}

File: sageattention/ops.py
"""
Copyright (c) 2024 by SageAttention team.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

---

FakeTensor kernels for the torch.ops.sageattention attention ops (needed for
torch.compile). The mutating quantization ops return () and need no fake.
"""

import torch

from ._layout import _seq_nh_dims


def _fwd_fake(
    query,
    key,
    value,
    query_scale,
    key_scale,
    value_scale=None,
    value_mean=None,
    *,
    tensor_layout="HND",
    qk_quant_gran="per_thread",
    pv_accum_dtype="fp32",
    v_layout="mma_k16",
    is_causal=False,
    sm_scale=1.0,
    return_lse=False,
    out_dtype=torch.float16,
    backend=None,
):
    out = torch.empty(query.shape, dtype=out_dtype, device=query.device)
    if not return_lse:
        return out, None
    seq_dim, nh_dim = _seq_nh_dims(tensor_layout)
    qo_len, num_qo_heads = query.size(seq_dim), query.size(nh_dim)
    lse = torch.empty(
        (query.size(0), num_qo_heads, qo_len), dtype=torch.float32, device=query.device
    )
    return out, lse


def _fwd_varlen_fake(
    query,
    key,
    value,
    query_scale,
    key_scale,
    cu_seqlens_q,
    cu_seqlens_k,
    value_scale=None,
    value_mean=None,
    *,
    max_seqlen_q,
    max_seqlen_k,
    qk_quant_gran="per_thread",
    pv_accum_dtype="fp32",
    v_layout="mma_k16",
    is_causal=False,
    sm_scale=1.0,
    return_lse=False,
    out_dtype=torch.float16,
):
    # packed [total_tokens, heads, head_dim]; the lse is head-major over the
    # same packed token axis. max_seqlen_* are never read: they size grid.x
    # only, which is what keeps them out of the traced shape environment.
    out = torch.empty(query.shape, dtype=out_dtype, device=query.device)
    if not return_lse:
        return out, None
    lse = torch.empty(
        (query.size(1), query.size(0)), dtype=torch.float32, device=query.device
    )
    return out, lse


def _ceil_div(a, b):
    return (a + b - 1) // b  # SymInt-friendly (no float round trip)


def _quant_qk_fake(
    query,
    key,
    key_mean=None,
    *,
    tensor_layout="HND",
    qk_quant_gran="per_thread",
    blk_q=128,
    warp_q=32,
    blk_k=64,
    warp_k=64,
):
    seq_dim, nh_dim = _seq_nh_dims(tensor_layout)
    qo_len, kv_len = query.size(seq_dim), key.size(seq_dim)
    if qk_quant_gran == "per_warp":
        qs = _ceil_div(qo_len, blk_q) * (blk_q // warp_q)
        ks = _ceil_div(kv_len, blk_k)
    else:
        qs = _ceil_div(qo_len, blk_q) * (blk_q // warp_q) * 8
        ks = _ceil_div(kv_len, blk_k) * (blk_k // warp_k) * 4
    dev = query.device
    return (
        torch.empty(query.shape, dtype=torch.int8, device=dev),
        torch.empty((query.size(0), query.size(nh_dim), qs), dtype=torch.float32, device=dev),
        torch.empty(key.shape, dtype=torch.int8, device=dev),
        torch.empty((key.size(0), key.size(nh_dim), ks), dtype=torch.float32, device=dev),
    )


def _blk_total(total_tokens, batch_size, cta_tokens):
    """csrc/sageattn/varlen.h blk_total: the packed block count of the whole
    batch. It reads (total_tokens, batch_size) only -- both static shapes --
    which is why no varlen fake kernel needs an unbacked SymInt."""
    return total_tokens // cta_tokens + batch_size


def _quant_qk_varlen_fake(
    query,
    key,
    cu_seqlens_q,
    cu_seqlens_k,
    key_mean=None,
    *,
    max_seqlen_q,
    max_seqlen_k,
    qk_quant_gran="per_thread",
    blk_q=128,
    warp_q=32,
    blk_k=64,
    warp_k=64,
):
    b = cu_seqlens_q.size(0) - 1
    q_blocks = _blk_total(query.size(0), b, blk_q)
    k_blocks = _blk_total(key.size(0), b, blk_k)
    if qk_quant_gran == "per_warp":
        qs = q_blocks * (blk_q // warp_q)
        ks = k_blocks
    else:
        qs = q_blocks * (blk_q // warp_q) * 8
        ks = k_blocks * (blk_k // warp_k) * 4
    dev = query.device
    return (
        torch.empty(query.shape, dtype=torch.int8, device=dev),
        torch.empty((query.size(1), qs), dtype=torch.float32, device=dev),
        torch.empty(key.shape, dtype=torch.int8, device=dev),
        torch.empty((key.size(1), ks), dtype=torch.float32, device=dev),
    )


def _segment_mean_varlen_fake(input, cu_seqlens, *, max_seqlen):
    # [batch_size, heads, head_dim]: batch_size is cu_seqlens.size(0) - 1, a
    # static shape. max_seqlen sizes the grid only, so it stays unread here.
    b = cu_seqlens.size(0) - 1
    return torch.empty((b, input.size(1), input.size(2)), dtype=input.dtype, device=input.device)


def _quant_v_fp8_fake(
    value,
    *,
    tensor_layout="HND",
    v_layout="mma_k16",
    scale_max=448.0,
    smooth_v=False,
    pad_multiple=64,
):
    seq_dim, nh_dim = _seq_nh_dims(tensor_layout)
    b, kv_len, h, d = (value.size(0), value.size(seq_dim), value.size(nh_dim), value.size(-1))
    padded = _ceil_div(kv_len, pad_multiple) * pad_multiple
    shape = (b, h, d, padded) if tensor_layout == "HND" else (b, d, h, padded)
    dev = value.device
    v_fp8 = torch.empty(shape, dtype=torch.float8_e4m3fn, device=dev)
    v_scale = torch.empty((b, h, d), dtype=torch.float32, device=dev)
    vm = torch.empty((b, h, d), dtype=torch.float32, device=dev) if smooth_v else None
    return v_fp8, v_scale, vm


def _quant_v_fp8_varlen_fake(
    value,
    cu_seqlens_k,
    *,
    max_seqlen_k,
    v_layout="mma_k16",
    scale_max=448.0,
    smooth_v=False,
    pad_multiple=64,
):
    # the padded V^T axis is blk_total blocks of pad_multiple tokens, which
    # reads (total_tokens, batch_size) only -- both static shapes
    b = cu_seqlens_k.size(0) - 1
    h, d = value.size(1), value.size(2)
    padded = _blk_total(value.size(0), b, pad_multiple) * pad_multiple
    dev = value.device
    v_fp8 = torch.empty((h, d, padded), dtype=torch.float8_e4m3fn, device=dev)
    v_scale = torch.empty((b, h, d), dtype=torch.float32, device=dev)
    vm = torch.empty((b, h, d), dtype=torch.float32, device=dev) if smooth_v else None
    return v_fp8, v_scale, vm


def _sub_mean_v_fake(value, *, tensor_layout="HND"):
    seq_dim, _ = _seq_nh_dims(tensor_layout)
    vm_shape = list(value.shape)
    del vm_shape[seq_dim]
    return (
        torch.empty(value.shape, dtype=torch.float16, device=value.device),
        torch.empty(vm_shape, dtype=value.dtype, device=value.device),
    )


def _register() -> None:
    torch.library.register_fake("sageattention::fwd")(_fwd_fake)
    torch.library.register_fake("sageattention::fwd_varlen")(_fwd_varlen_fake)
    torch.library.register_fake("sageattention::quant_qk")(_quant_qk_fake)
    torch.library.register_fake("sageattention::quant_qk_varlen")(_quant_qk_varlen_fake)
    torch.library.register_fake("sageattention::segment_mean_varlen")(_segment_mean_varlen_fake)
    torch.library.register_fake("sageattention::quant_v_fp8")(_quant_v_fp8_fake)
    torch.library.register_fake("sageattention::quant_v_fp8_varlen")(_quant_v_fp8_varlen_fake)
    torch.library.register_fake("sageattention::sub_mean_v")(_sub_mean_v_fake)


_register()
